### PR TITLE
Cleanup Azure resource groups

### PR DESCRIPTION
**Pull request by adam-stokes**
Utilize a predefined AZ resource group prefix, we query the available resource
groups utilizing JMESPATH pulling out the group names. Finally passing that
through powershell's ForEach block to remove those remaining groups.

Related https://github.com/elastic/observability-robots/issues/497

Signed-off-by: Adam Stokes <51892+adam-stokes@users.noreply.github.com>

### CODE REVIEW

**Comment by adam-stokes** [comment]
@russcam @elastic/observablt-robots Looking for feedback on this approach, if worthwhile I may generalize this into a helper function to be put in our pipeline library.

File: Jenkinsfile
Comment (russcam) [inline Jenkinsfile:19]: This would need to include some unique identifier for a specific build run, so that the cleanup step only deletes resource groups created in this run. Is the Jenkins `BUILD_ID` unique per run? Could that be included?

File: Jenkinsfile
Comment (russcam) [inline Jenkinsfile:518]: Before `az group list` can run, 

1. [the Service Principal credentials in `.credentials.json` will need to be used to log in](https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli#sign-in-with-a-service-principal). Something like the following (this needs checking/testing to ensure sensitive details are not captured in logs or displayed in CI)

```sh
az login --service-principal --username <clientId> --password <clientSecret> --tenant <tenantId>
```

2. Since a tenant can contain multiple subscriptions, the active subscription to work with should be set as the one from `.credentials.json`

```sh
az account set --subscription <subscriptionId>
```

Now, `az group list` is good to go

Comment (russcam) [reply Jenkinsfile:518]:
```
jq .[].name
```
Could avoid needing to install `jq`, if it's just for this, using PowerShell's [`ConvertFrom-Json`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-json?view=powershell-7.1). Also would want [`--nowait` and `--yes` when deleting groups](https://docs.microsoft.com/en-us/cli/azure/group?view=azure-cli-latest#az_group_delete). Something like

```powershell
.. | ConvertFrom-Json | ForEach { az group delete --name $_.name --no-wait --yes  }
```

**Review by russcam — CHANGES_REQUESTED** [review]
I left some comments for consideration

File: Jenkinsfile
Comment (russcam) [reply Jenkinsfile:518]: Would need a bash/sh variant too, for Linux build hosts?

File: Jenkinsfile
Comment (adam-stokes) [inline Jenkinsfile:521]: @russcam I think the cleanup always runs on the parent node which seems to be linux (ubuntu) is there a situation where this may not be the case? If not, we could probably remove this windows portion

Comment (russcam) [reply Jenkinsfile:521]: I'm not super familiar with the Jenkins setup so will defer to @elastic/observablt-robots. If it helps, it doesn't matter from what node/host the cleanup task runs from, so long as the input parameters are correct.

File: Jenkinsfile
Comment (russcam) [reply Jenkinsfile:563]: looks like withAzureCredentials will get the right set of field names, though it writes them to file. Ideally want a step that gets the field names from withAzureCredentials, but sets the values as environment variables 🙂 Parsing the file with jq would work though

Comment (adam-stokes) [reply Jenkinsfile:563]: @v1v I looked at `withAzureEnv` however it looks like it's pulling different key names than what's provided in that closure.

@russcam I decided against `withAzureCredentials` as to not have to worry about masking the output from `jq` or reading the file in general. We already had closures for handling this task via vault so this was more straightforward.

**Comment by cachedout** [comment]
@adam-stokes Just trying to move this one along before we hit the end of our sprint. :)

This looks like it's set to WIP but [your most recent comment suggested that you're waiting for reviews](https://github.com/elastic/observability-robots/issues/497#issuecomment-897633230). If this is ready to go, it's probably best to remove the WIP status and see if we can get the reviews needed to get this moving forward. Thanks.

**Review by mdelapenya — APPROVED** [review]

File: Jenkinsfile
Comment (adam-stokes) [reply Jenkinsfile:566]: Hey @russcam I believe the host is always ephemeral as it's booting a vm with `immutable`. But just to be extra cautious we can remove it after cleanup occurs

**Comment by adam-stokes** [comment]
> The `AZURE_RESOURCE_GROUP_PREFIX` environment variable needs to be used on the .NET side, when spinning up resources with Terraform. Would you like me to push a commit to this PR to update this?

That would be awesome @russcam if you aren't able to get to it during your work hours I can tackle it in the morning as well.

Thanks!

**Comment by russcam** [comment]
I've added a commit to use the `AZURE_RESOURCE_GROUP_PREFIX` environment variable

**Comment by adam-stokes** [comment]
@russcam Ok everything is ready to go, the last hiccup is the resource name itself giving an error:

```
09:31:33  + docker run --rm -i -v /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1377/apm-agent-dotnet/.azure:/root/.azure mcr.microsoft.com/azure-cli:latest az group list --query [?name | starts_with(@,'ci-dotnet-18')] --out json
09:31:34  + pwd
09:31:34  + docker run --rm -i -v /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1377/apm-agent-dotnet/.azure:/root/.azure mcr.microsoft.com/azure-cli:latest az group delete --name "ci-dotnet-18-apm-ci-immutable-ubuntu-1804-1629463347797644918-cosmosdb-test" --no-wait --yes
09:31:35  ERROR: Parameter 'resource_group_name' must conform to the following pattern: '^[-\\w\\._\\(\\)]+$'.
```

The only reference I could find was https://github.com/Azure/azure-cli/issues/1284#issuecomment-476749240

**Comment by russcam** [comment]
The resource name looks OK, [here are the rules](https://docs.microsoft.com/en-us/rest/api/resources/resource-groups/create-or-update#uri-parameters)

> Can include alphanumeric, underscore, parentheses, hyphen, period (except at end), and Unicode characters that match the allowed characters.
> Regex pattern: `^[-\w\._\(\)]+$`

**Comment by adam-stokes** [comment]
> The resource name looks OK, [here are the rules](https://docs.microsoft.com/en-us/rest/api/resources/resource-groups/create-or-update#uri-parameters)
> 
> > Can include alphanumeric, underscore, parentheses, hyphen, period (except at end), and Unicode characters that match the allowed characters.
> > Regex pattern: `^[-\w\._\(\)]+$`

I think it was the `"` as jq without `--raw-output` will quote the strings when printing to stdout. This most recent change should address that

**Pull request merged**
